common/base_models.py

```diff
@@ -0,0 +1,111 @@
+import sys
+import uuid
+
+from model_utils.models import TimeStampedModel, SoftDeletableModel
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from django.db import models
+from simple_history.models import HistoricalRecords
+
+TEST = 'test' in sys.argv
+
+
+class ImmutableBaseModel(SoftDeletableModel, TimeStampedModel):
+    """Used as a base class for models that are immutable and additive over time.
+
+    A subclass is expected to always provide a user model for creation. If subclasses are
+    to be saved outside of an authenticated user session, a common 'automated user' should be applied,
+    allowing audit to confirm that an automated task generated the record.
+
+    As a default, it provides attributes:
+
+      id -- UUID
+      creator -- User that created the record
+      created -- Timestamp the model was created
+      is_removed -- True if the instance is to be considered deleted and ignored in queries
+    """
+
+    def __init__(self, *args, **kwargs):
+        if not hasattr(self, '_current_user'):
+            self._current_user = None
+        super().__init__(*args, **kwargs)
+
+    class Meta:
+        abstract = True
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    creator = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.PROTECT,
+                                null=False, related_name='%(class)s_creator')
+
+    @property
+    def current_user(self):
+        return self._current_user
+
+    @current_user.setter
```

Comment:
Why go through all the setting of a `_current_user` proxy on the model instance? Is that to set the `creator` field in the end? 

If that's the goal a more common practice is just to leave all of that out and use a field (or better, set that field through a mixin) on the serializer. That would be something along the lines of `creator = serializers.HiddenField(default=CurrentUserDefault())`. If the user is not authenticated, it will use `AnonymousUser`.

Reply:
The reason is to ensure that when a model is saved, it automatically forces the appropriate creator / modifier fields to be set. This logic is part of the model this way, not the responsibility for consumers to remember whether they are creating or updating and which field to set when (or both, maybe?!). Of course, the serializer can do it, as can the model admin, but then model tests become responsible for setting the right thing too. The logic belongs, IMO in the model.

I have learnt from working on a project that needs heavy compliance and auditability that it is better to not leave this to chance, and just ensure that all models that have to be persisted have a current_user set, consistently. Then it just happens. If you forget, then it fails. 

You'll see from the tests that exist right now that all that has to be set in a `objects.create` is the extra property `current_user`. As a consumer I don't have to think about it, just do it. Of course, it does lead to some extra complexity in the base model, but I'm alright with that, unless it really makes you uncomfortable.